### PR TITLE
Remove empty references from templates

### PR DESCRIPTION
**Pull request by rafaelgallani**
Changes the template to remove all empty interfaces, traits, and namespaces. 
We consider these as "empty" when there's no method exposed for the Twig writer used in phpdocumentor.

It removes the sections and marks docs with a comment, so the script that runs after can clean up the marked files.

### CODE REVIEW

**Comment by ramynasr** [comment]
* What would be a good way to test how this works?
* Are there examples of files that would exist before that change and would be deleted after?
* If we remove files, would their links also be removed? Or would we end up with broken links?

**Comment by rafaelgallani** [comment]
@ramynasr 
Thanks for the review. I added almost no context to the PR. Going through the points a little better:

> * What would be a good way to test how this works?

You can try adding a class that implements an interface marked with `@ignore`.
The interface won't be generated, but the class will still reference it (404). 

This PR will remove the (inaccessible) interface from the docs.

> * Are there examples of files that would exist before that change and would be deleted after?

There are a few examples here - this PR uses this branch: 
https://github.com/Automattic/domain-services-client/pull/153/commits/25cea37c87496e2b0ea61d18afdab4b6515c897b

You can see that only empty namespaces files will be removed - interfaces that have methods will still be generated, and the ones that are ignored with `@ignore` already weren't being generated.

> * If we remove files, would their links also be removed? Or would we end up with broken links?

The links must be removed manually, which is what this PR does. 
Currently, if we don't generate an interface/trait docs, we end up with a 404 - as in the first bullet point example, a class with an interface marked with `@ignore`.

These links were removed here (the `if` condition):

https://github.com/Automattic/phpdoc-markdown/blob/2fd3463a35b6cb2c5c65df29704588dd226e29f5/data/templates/markdown/class.md.twig#L69

And here:

https://github.com/Automattic/phpdoc-markdown/blob/2fd3463a35b6cb2c5c65df29704588dd226e29f5/data/templates/markdown/class.md.twig#L75

**Review by leonardost — APPROVED** [review]
LGTM! Did some testing by changing the dependency in `domain-services-client` to `"automattic/phpdoc-markdown": "dev-add/remove-empty-references"` and this seems to work as intended, no more references to `@internal` entities are shown in the docs. 🚀

File: data/templates/markdown/namespace.md.twig [hunk 1]
```diff
@@ -1,4 +1,7 @@
 {% extends './layout.md.twig' %}
 {% block content %}
 {{ block('table_of_contents') -}}
+{% if ( node.children|length == 0 ) and ( node.traits|length == 0 ) and ( node.interfaces|length == 0 ) and ( node.classes|length == 0 ) and ( node.constants|length == 0 ) and ( node.functions|length == 0 ) %}
```
Comment (ramynasr) [inline data/templates/markdown/namespace.md.twig]: If we do something like:

```
{% if ( node.children|length != 0 ) or ( node.traits|length != 0 ) or ( node.interfaces|length != 0 ) or ( node.classes|length != 0 ) or ( node.constants|length != 0 ) or ( node.functions|length != 0 )  or ( node.methods|length != 0 )%}
# {% block title %}{{ macros.buildNamespaceBreadcrumb(node)|spaceless|replace({"\n": "", "\r": "", "\t": ""}) }}{% endblock %}

{% block content %}{% endblock %}
{% block footer %}{% endblock %}
{% endif %}
```

in `layout.html.twig`. then in `html-to-md.php` we do something like:

```
    $content = file_get_contents($file);

	if (empty($content)) {
		unlink( $file);
		echo 'DELETED' . PHP_EOL;
		continue;
	}
```

I believe we won't need to have a specific string to test for in the script and we can just check if the file is empty or not.

File: data/templates/markdown/elements/table-of-contents.md.twig [hunk 2]
```diff
@@ -1,10 +1,18 @@
 {% block table_of_contents %}
 {% set url_base = (url_base) ? url_base : '../' %}
+{% set non_empty_namespaces = [] %}
 {% if node.children|length > 0 %}
+{% for namespace in node.children|sort_asc %}
+{% if ( namespace.children|length > 0 ) or ( namespace.traits|length > 0 ) or ( namespace.interfaces|length > 0 ) or ( namespace.classes|length > 0 ) or ( namespace.constants|length > 0 ) or ( namespace.functions|length > 0 ) %}
```
Comment (ramynasr) [inline data/templates/markdown/elements/table-of-contents.md.twig]: I believe for `constants`, `properties` and `methods`.. we need to use the methods `constants(namespace)`, `properties(namespace)` and `methods(namespace)` respectively.

`namespace.methods` would not include inherited methods for example, while `methods(namespace)` would.

File: data/templates/markdown/class.md.twig [hunk 3]
```diff
@@ -66,13 +66,13 @@
...
 * File: [{{ node.path }}]({{ '../' ~ path(node.file) }})
-{% for interface in node.interfaces %}
+{% for interface in node.interfaces if interface.methods|length > 0 %}
 {% if loop.first %}
 * Implements:
 {% endif %}
   * [{{ interface }}]({{ link(interface) }})
 {% endfor %}
-{% for trait in node.usedTraits %}
+{% for trait in node.usedTraits if trait.methods|length > 0 %}
```
Comment (ramynasr) [inline data/templates/markdown/class.md.twig]: I believe we need to use `methods(trait)` to attribute for inheritance. Also traits can have properties (constants too starting PHP `8.2`) and no methods, so maybe we need to check for that too?

File: data/templates/markdown/class.md.twig [hunk 4]
```diff
@@ -66,13 +66,13 @@
 ### Details
 
 * File: [{{ node.path }}]({{ '../' ~ path(node.file) }})
-{% for interface in node.interfaces %}
+{% for interface in node.interfaces if interface.methods|length > 0 %}
```
Comment (ramynasr) [inline data/templates/markdown/class.md.twig]: Same comment for traits below. Also, interfaces can have constants.

**Review by ramynasr — CHANGES_REQUESTED** [review]

**Comment by ramynasr** [comment]
Thanks for going through this @rafaelgalani! looks great! I have some concerns about inheritance that I only know about because I was recently working on this. Please let me know if that makes sense and if you have any questions we can pair on it.

File: data/templates/markdown/class.md.twig (hunk 3 above)
Comment (rafaelgallani) [reply data/templates/markdown/class.md.twig]: I checked that phpdocumentor v3 (our fork's template version) supports constants for traits - see here:
https://github.dev/phpDocumentor/phpDocumentor/blob/eec5cc7e0f384c603116dc6d759886bf1120048c/src/phpDocumentor/Descriptor/TraitDescriptor.php#L36

And here (trait):
https://github.dev/phpDocumentor/phpDocumentor/blob/eec5cc7e0f384c603116dc6d759886bf1120048c/src/phpDocumentor/Descriptor/Traits/HasConstants.php#L23-L62

So I added the check.

---

But for some reason (which I don't know), when I inspect the `TraitDescriptor.php` in the phar used in [DSAPI](https://github.com/Automattic/domain-services-client), it doesn't include these methods. I'm not familiar with phar files, but this might be something we want to check later. 

FWIW, I also [added a check for the method used internally within the Twig extension](https://github.com/Automattic/phpdoc-markdown/pull/2/files#diff-504b094b23ce85994c6ff1838c633ee4adf2aa13c7a501b71902ff5cc4e2caa3R75) anyway.

File: data/templates/markdown/namespace.md.twig (hunk 1 above)
Comment (rafaelgallani) [reply data/templates/markdown/namespace.md.twig]: I made it slightly different, but the general idea was the same - check for an empty file instead of a specific string (much better).
I had to do it this different way because otherwise, we'd still have to check for a `# ` string - this would be the file's content when "empty".

I also like it a little better because we can replace the heading - if we want to.

File: bin/html-to-md.php [hunk 5]
```diff
@@ -67,5 +69,15 @@ function get_target_dir(): string {
     file_put_contents($file, $content);
     $mdFilePath = preg_replace('/\.html$/', '.md', $file);
     rename($file, $mdFilePath);
+	$newFiles[] = $mdFilePath;
 	echo 'DONE' . PHP_EOL;
 }
+
+foreach ($newFiles as $file) {
```
Comment (ramynasr) [inline bin/html-to-md.php]: Do we need a new loop where we have to loop and read all content again? What if we add that check to like `62`?

**Review by ramynasr — APPROVED** [review]
Thank you for taking care of all the changes! LGTM
